File: lib/departures.ts
/**
 * The scheduled departure board for one boarding point.
 *
 * This is a different capability from matching a vehicle to our motion model, and it is built from
 * a different thing: the operator's own registered timetable, which the pipeline already parses
 * (`pipeline/departures.py`). A row here is a *scheduled* departure and says so. Nothing on this
 * board is a prediction, and no row borrows a time from a bus: where a tracked vehicle can be tied
 * to a scheduled journey it is named beside that row, and where it cannot the row stands alone.
 *
 * Times are seconds from local midnight on the service day, so a journey timed at 00:12 the next
 * morning is 86 520 rather than 720 and stays in order. Each is turned into a real instant against
 * Europe/London, so a countdown crosses midnight and both clock changes without arithmetic on
 * wall-clock strings.
 */
import {z} from 'zod';
import {type OperatingRule,londonDate,ruleApplies} from '@/lib/service-days';

const ruleSchema:z.ZodType<OperatingRule>=z.object({
 days:z.array(z.number()).optional(),holidaysOnly:z.boolean().optional(),
 alsoOn:z.array(z.tuple([z.string(),z.string()])).optional(),
 notOn:z.array(z.tuple([z.string(),z.string()])).optional(),
 serviced:z.array(z.object({mode:z.enum(['only','except']),kind:z.string(),
  organisations:z.array(z.string()),ranges:z.array(z.tuple([z.string(),z.string()]))})).optional(),
 bankHolidays:z.string().optional(),
}).passthrough() as unknown as z.ZodType<OperatingRule>;

const serviceSchema=z.object({
 patternId:z.string(),operator:z.string().nullable(),line:z.string(),
 direction:z.string().nullable(),destination:z.string().nullable(),sequence:z.number(),
 rules:z.array(z.number()),offsets:z.array(z.number().nullable()),
 runs:z.array(z.array(z.number())),
 timetable:z.object({file:z.string().nullable(),datasetSha256:z.string().nullable(),
  validFrom:z.string().nullable(),validTo:z.string().nullable()}).partial().optional(),
});
const boardSchema=z.object({schemaVersion:z.number(),stop:z.string(),generatedAt:z.string(),
 services:z.array(serviceSchema)});
const rulesSchema=z.object({schemaVersion:z.number(),rules:z.array(ruleSchema)});

export type StopBoardService=z.infer<typeof serviceSchema>;
export type StopDepartures={stop:string;generatedAt:string;services:StopBoardService[]};

/** One scheduled departure from this stop, on a named journey of a named pattern. */
export type ScheduledDeparture={
 /** Local midnight of the service day it belongs to, plus its own seconds. */
 atMs:number;
 /** The operator's scheduled departure from the pattern's *first* stop, HH:MM:SS local: what the
  *  feed reports as OriginAimedDepartureTime, and the only key a vehicle may be joined on. */
 originLocal:string;
 originSeconds:number;
 /** Seconds from that service day's local midnight at *this* stop: past 86 400 for a journey
  *  timed after midnight, which is how it keeps its place in the order. */
 stopSeconds:number;
 serviceDay:string;
 line:string;destination:string;direction:string|null;operator:string|null;patternId:string;
 /** How many journeys of this pattern leave the origin at that time: more than one and no single
  *  vehicle can be named for this row. */
 sharedDeparture:number;
 /** False when the journey carried no operating profile we could read: it may not run today. */
 dayKnown:boolean;
};

const SECONDS_IN_DAY=86_400;
const LONDON=new Intl.DateTimeFormat('en-GB',{timeZone:'Europe/London',hour12:false,
 year:'numeric',month:'2-digit',day:'2-digit',hour:'2-digit',minute:'2-digit',second:'2-digit'});

/** Europe/London's offset from UTC at that instant, in milliseconds. */
function offsetAt(ms:number):number{
 const parts=Object.fromEntries(LONDON.formatToParts(ms).map(part=>[part.type,part.value]));
 const asUtc=Date.UTC(Number(parts.year),Number(parts.month)-1,Number(parts.day),
  Number(parts.hour)%24,Number(parts.minute),Number(parts.second));
 return asUtc-Math.floor(ms/1000)*1000;
}

/**
 * The instant of `seconds` past local midnight on `iso`, in Europe/London.
 *
 * Taken in two passes, because the offset that applies is the one at the answer, not the one at
 * the guess: on the last Sunday in October 01:30 happens twice and on the last Sunday in March it
 * does not happen at all, and a timetable states wall-clock times either way.
 */
export function londonInstant(iso:string,seconds:number):number{
 const [year,month,day]=iso.split('-').map(Number);
 const wall=Date.UTC(year,(month??1)-1,day??1)+seconds*1000;
 return wall-offsetAt(wall-offsetAt(wall));
}

/** Local midnight of the service day a moment belongs to: a timetable's day, not the clock's. */
export const serviceDayOf=(nowMs:number)=>londonDate(nowMs);

/** A day either side of an ISO date: `-1` for the journeys timed past midnight that still belong
 *  to yesterday, `+1` for tomorrow morning's, which is what is next at eleven at night. */
export function shiftDay(iso:string,days:number):string{
 const [year,month,day]=iso.split('-').map(Number);
 return new Date(Date.UTC(year,(month??1)-1,(day??1)+days)).toISOString().slice(0,10);
}
export const previousDay=(iso:string)=>shiftDay(iso,-1);

const label=(value:string|null|undefined)=>(value??'').replace(/_/g,' ').trim();

/** Every departure of one service on one service day, as instants. */
function departuresOf(service:StopBoardService,rules:OperatingRule[],day:string):ScheduledDeparture[]{
 const out:ScheduledDeparture[]=[];
 const shared=new Map<number,number>();
 for(const run of service.runs){
  let at=run[2];
  for(let i=2;i<run.length;i++){
   if(i>2)at+=run[i];
   shared.set(at,(shared.get(at)??0)+1);
  }
 }
 for(const run of service.runs){
  const ruleIndex=run[0],timing=run[1];
  const rule=ruleIndex>=0?rules[service.rules[ruleIndex]??-1]:undefined;
  // A journey whose own profile we could not read is neither claimed to run nor hidden: it is
  // carried with `dayKnown` false, and the board says the day is unconfirmed rather than guessing.
  const dayKnown=rule!==undefined;
  if(rule&&!ruleApplies(rule,day))continue;
  const offset=service.offsets[timing];
  if(offset===null||offset===undefined)continue;
  let origin=run[2];
  for(let i=2;i<run.length;i++){
   if(i>2)origin+=run[i];
   const seconds=origin+offset;
   const hours=Math.floor(origin/3600),minutes=Math.floor((origin%3600)/60),secs=origin%60;
   out.push({
    atMs:londonInstant(day,seconds),
    originLocal:`${String(hours%24).padStart(2,'0')}:${String(minutes).padStart(2,'0')}:${String(secs).padStart(2,'0')}`,
    originSeconds:origin,stopSeconds:seconds,serviceDay:day,
    line:service.line,destination:label(service.destination),direction:service.direction,
    operator:service.operator,patternId:service.patternId,
    sharedDeparture:shared.get(origin)??1,dayKnown,
   });
  }
 }
 return out;
}

/**
 * The next `limit` scheduled departures from this stop at `nowMs`.
 *
 * Two service days are read: today's, and yesterday's for the journeys timed past midnight that
 * are still running. A departure already gone is dropped by `graceSeconds`, not by nothing: a bus
 * timed at 21:14 is still worth showing at 21:14:30.
 */
export function nextDepartures(board:StopDepartures|null,rules:OperatingRule[]|null,nowMs:number,
                               {limit=8,graceSeconds=60,withinMinutes=180}={}):ScheduledDeparture[]{
 if(!board||!rules)return [];
 const today=serviceDayOf(nowMs),yesterday=previousDay(today),tomorrow=shiftDay(today,1);
 const all:ScheduledDeparture[]=[];
 for(const service of board.services){
  all.push(...departuresOf(service,rules,today));
  // Only the ones that ran past midnight; the rest of yesterday is over. Judged on the time at
  // *this* stop, which is the departure's own — two journeys of one pattern can be timed
  // differently, so the pattern's first offset is not theirs.
  for(const departure of departuresOf(service,rules,yesterday))
   if(departure.stopSeconds>=SECONDS_IN_DAY)all.push(departure);
  // And tomorrow's, because at eleven at night the next bus from this stop is tomorrow's, and a
  // board that simply went blank would be saying something it does not know.
  all.push(...departuresOf(service,rules,tomorrow));
 }
 const from=nowMs-graceSeconds*1000,to=nowMs+withinMinutes*60_000;
 return all.filter(d=>d.atMs>=from&&d.atMs<=to).sort((a,b)=>a.atMs-b.atMs).slice(0,limit);
}

/** Minutes until a departure, from the instants themselves: never from a clock-face subtraction. */
export const minutesUntil=(departure:{atMs:number},nowMs:number)=>
 Math.round((departure.atMs-nowMs)/60_000);

export function countdownWords(departure:{atMs:number},nowMs:number):string{
 const minutes=minutesUntil(departure,nowMs);
 if(minutes<=0)return 'due';
 if(minutes===1)return '1 min';
 if(minutes<60)return `${minutes} min`;
 const hours=Math.floor(minutes/60);
 return `${hours} h ${minutes-hours*60} min`;
}

export const clockWords=(departure:{atMs:number})=>new Intl.DateTimeFormat('en-GB',
 {timeZone:'Europe/London',hour:'2-digit',minute:'2-digit',hour12:false}).format(departure.atMs);

const boardCache=new Map<string,Promise<StopDepartures|null>>();
let rulesRequest:Promise<OperatingRule[]|null>|null=null;

/**
 * A fetch that is certain to settle. A board that says "reading the timetable" has promised
 * something about the near future, and a promise that a hung request can hold for ever is the
 * fault this milestone fixed in the front view's "checking" (docs/REVIEW.md).
 */
async function fetchWithin(url:string,ms=8000):Promise<Response|null>{
 const abort=new AbortController();
 const timer=setTimeout(()=>abort.abort(),ms);
 try{return await fetch(url,{signal:abort.signal})}catch{return null}finally{clearTimeout(timer)}
}

/** The shared operating rules every board's services point at: one fetch, kept for the visit. */
export function loadDepartureRules(base='/data'):Promise<OperatingRule[]|null>{
 rulesRequest??=(async()=>{
  const response=await fetchWithin(`${base}/departure-rules.json`);
  if(!response?.ok)return null;
  const parsed=rulesSchema.safeParse(await response.json().catch(()=>null));
  return parsed.success?parsed.data.rules:null;
 })();
 return rulesRequest;
}

/** One stop's board. A stop with no board published is null, never an empty board: they differ. */
export function loadStopDepartures(atco:string|null|undefined,base='/data'):Promise<StopDepartures|null>{
 if(!atco)return Promise.resolve(null);
 const cached=boardCache.get(atco);
 if(cached)return cached;
 const request=(async():Promise<StopDepartures|null>=>{
  const response=await fetchWithin(`${base}/departures/${encodeURIComponent(atco)}.json`);
  if(!response?.ok)return null;
  const parsed=boardSchema.safeParse(await response.json().catch(()=>null));
  return parsed.success?{stop:parsed.data.stop,generatedAt:parsed.data.generatedAt,
   services:parsed.data.services}:null;
 })();
 boardCache.set(atco,request);
 return request;
}

/** For tests: forget what has been fetched. */
export function resetDepartureCache(){boardCache.clear();rulesRequest=null}

File: pipeline/departures.py
"""Scheduled departure boards, one file per boarding point.

    .venv/bin/python -m pipeline.departures publish

A stop's board is a different capability from matching a vehicle to our motion model, and it is
built from a different thing: the operator's own registered timetable, which we already hold and
parse. For every pattern that calls at a stop, every journey's departure from the pattern's first
stop is carried forward by that journey's declared running time to this stop. That is a scheduled
time, said as one — never a prediction, never adjusted for where a bus is.

What it refuses, rather than filling in:
  - a pattern whose stop has no declared running time (`seconds_from_start` null) publishes no
    time at that stop, because the file did not state one and a zero would be an invention;
  - a departure whose journey carried no operating profile we could read (rule -1) is kept and
    marked, so a board can say it may not run today rather than silently listing or dropping it;
  - a stop a pattern ends at is not a departure from it: nobody boards there.

Times are seconds from that service day's midnight, local. A journey timed past midnight keeps a
value over 86400 rather than wrapping, so "00:12 tomorrow" stays after "23:55 tonight" in order
and can be turned into a real timestamp by the page without guessing which day it meant.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

from .core import atomic_json, utc_now
from .warehouse import DEFAULT_DB, connect

TARGET = Path('public/data/departures')
# A day's own seconds: a departure at 25:10:00 in the file is 12 minutes past midnight, next day.
SECONDS_IN_DAY = 86_400


def _total(run):
    """The last absolute time in a delta-encoded run: [rule, timing, first, gap, gap, ...]."""
    return sum(run[2:])


def _seconds(text):
    try:
        parts = [int(p) for p in str(text).split(':')]
    except (TypeError, ValueError):
        return None
    if len(parts) < 2:
        return None
    hours, minutes = parts[0], parts[1]
    seconds = parts[2] if len(parts) > 2 else 0
    return hours * 3600 + minutes * 60 + seconds


def boards(con, only=None):
    """Every stop's scheduled departures, keyed by ATCO code.

    `only` keeps the boards to the stops we publish at all (the service area): a pattern runs well
    beyond it, and a board for a stop the page can never show is 40 MB nobody reads.
    """
    patterns = con.execute("""
        SELECT pattern_id, operator_code, line_name, direction, destination_display,
               operating_rules, departure_times, valid_from, valid_to, source_file, dataset_sha256
        FROM service_pattern""").fetchall()
    stops_by_pattern = {}
    for pattern_id, sequence, atco, _metres, seconds in con.execute("""
        SELECT pattern_id, sequence, atco_code, distance_from_start_m, seconds_from_start
        FROM service_pattern_stop ORDER BY pattern_id, sequence""").fetchall():
        stops_by_pattern.setdefault(pattern_id, []).append((sequence, atco, seconds))

    # One operating rule can carry a school calendar of forty date ranges, and the same rule is
    # shared by hundreds of patterns. Written into every service of every board it was 40 of the
    # 48 MB; written once and referenced by index it is a few kilobytes in one file.
    catalogue, index = [], {}

    def shared(rule):
        key = json.dumps(rule, sort_keys=True)
        if key not in index:
            index[key] = len(catalogue)
            catalogue.append(rule)
        return index[key]

    by_stop = {}
    for (pattern_id, operator, line, direction, destination, rules_json, departures_json,
         valid_from, valid_to, source_file, dataset) in patterns:
        stops = stops_by_pattern.get(pattern_id) or []
        if len(stops) < 2:
            continue
        info = json.loads(departures_json) if departures_json else {}
        timings, departures = info.get('timings') or [], info.get('departures') or []
        if not departures:
            continue
        rules = json.loads(rules_json) if rules_json else None
        last_index = stops[-1][0]
        for sequence, atco, own_seconds in stops:
            if sequence == last_index:
                continue                      # a pattern that ends here cannot be boarded here
            if only is not None and atco not in only:
                continue
            # Each journey's own declared running time from the origin to this stop, where the
            # file states one. The board keeps the *origin* departure, not the time at this stop,
            # because the origin departure is what the feed reports (OriginAimedDepartureTime) and
            # so the only thing a tracked vehicle can honestly be joined to a scheduled journey on.
            offsets = []
            for timing_index in range(max(1, len(timings))):
                row = timings[timing_index] if timing_index < len(timings) else []
                value = row[sequence] if sequence < len(row) else None
                offsets.append(own_seconds if value is None else int(value))
            times = []
            for row in departures:
                start = _seconds(row[0])
                if start is None:
                    continue
                timing = row[1] if len(row) > 1 else 0
                rule = row[2] if len(row) > 2 else -1
                if not 0 <= timing < len(offsets) or offsets[timing] is None:
                    continue                  # not declared anywhere: no time is published here
                times.append([start, int(timing), int(rule)])
            if not times:
                continue
            times.sort()
            # Grouped by the rule and timing the journey runs on, and delta-encoded from the first
            # departure in each group: a day's departures at a busy stop go from about 12 bytes
            # each to about 4. A run is [ruleIndex, timingIndex, firstOriginSeconds, gap, gap, …].
            runs = []
            for when, timing, rule in times:
                if runs and runs[-1][0] == rule and runs[-1][1] == timing:
                    runs[-1].append(when - _total(runs[-1]))
                else:
                    runs.append([rule, timing, when])
            by_stop.setdefault(atco, []).append({
                'patternId': pattern_id, 'operator': operator, 'line': line,
                'direction': direction, 'destination': destination, 'sequence': sequence,
                'rules': [shared(rule) for rule in (rules or [])], 'runs': runs,
                'offsets': offsets,
                'timetable': {'file': source_file, 'datasetSha256': dataset,
                              'validFrom': str(valid_from) if valid_from else None,
                              'validTo': str(valid_to) if valid_to else None},
            })
    return by_stop, catalogue


def publish(con, root=Path('.'), target=TARGET):
    directory = Path(root) / target
    directory.mkdir(parents=True, exist_ok=True)
    generated = utc_now()
    written, departures, existing = 0, 0, {p.name for p in directory.glob('*.json')}
    area = {row[0] for row in
            con.execute("SELECT atco_code FROM stop WHERE status = 'active'").fetchall()}
    by_stop, rule_catalogue = boards(con, only=area or None)
    for atco, services in by_stop.items():
        services.sort(key=lambda s: (s['line'], s['direction'], s['destination']))
        departures += sum(len(run) - 2 for s in services for run in s['runs'])
        atomic_json(directory / f'{atco}.json', {
            'schemaVersion': 1, 'stop': atco, 'generatedAt': generated,
            'secondsInDay': SECONDS_IN_DAY,
            'runsFormat': 'Each run is [ruleIndex, timingIndex, firstOriginSecondsFromLocalMidnight, '
                          'gap, gap, …]; add the gaps up for the rest. The scheduled time at this '
                          'stop is that origin departure plus services[].offsets[timingIndex]. '
                          'ruleIndex -1 means the journey carried no operating profile we could read.',
            'basis': 'Scheduled departures from the operators’ registered TransXChange timetables. '
                     'Seconds from local midnight on the service day; a time past 86400 is after '
                     'midnight. Not a prediction and not adjusted for where any bus is.',
            'attribution': 'Timetables: TfGM via the Bus Open Data Service (Open Government Licence v3.0).',
            'services': services})
        existing.discard(f'{atco}.json')
        written += 1
    # A stop that has lost its last service keeps no stale board.
    for stale in existing:
        (directory / stale).unlink(missing_ok=True)
    index = {'schemaVersion': 1, 'generatedAt': generated, 'stops': written,
             'departures': departures, 'removed': len(existing), 'rules': len(rule_catalogue),
             'basis': 'One board per boarding point under public/data/departures/<ATCO>.json. A '
                      'service’s `rules` are indices into public/data/departure-rules.json.'}
    atomic_json(Path(root) / 'public/data/departure-rules.json',
                {'schemaVersion': 1, 'generatedAt': generated, 'rules': rule_catalogue})
    atomic_json(Path(root) / 'public/data/departures.json', index)
    return index


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('command', choices=['publish'])
    parser.add_argument('--db', default=None)
    args = parser.parse_args(argv)
    con = connect(args.db or Path('.') / DEFAULT_DB)
    try:
        print(json.dumps(publish(con)))
    finally:
        con.close()
    return 0


if __name__ == '__main__':
    sys.exit(main())
